### DISCUSSION TITLE
Side effect that depends upon asynchronous value

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#41] [Q&A] [answered]
I have this code:

```dart
Future<ResourceCapsuleSideEffectController<bool>> bashrcPatchedRootUserResource(
  CapsuleHandle use,
) async {
  final getConnection = use(commons.connectionCapsule);

  return use.asResource<bool>(
    initialValue:
        await (await use(rootBashrcTaggedFileUtility)).isPatched(),
    getConnection: getConnection,
  );
}
```

I suppose we can not do `use.asResource` after doing `await` (twice).

Will be the following a good solution?
```dart
Future<ResourceCapsuleSideEffectController<bool>> bashrcPatchedRootUserResource(
  CapsuleHandle use,
) async {
  final useAsResource = use.asResource;

  final getConnection = use(commons.connectionCapsule);

  return useAsResource<bool>(
    initialValue: await (await use(rootBashrcTaggedFileUtility)).isPatched(),
    getConnection: getConnection,
  );
}
```

**Comment by GregoryConrad** [comment, marked as answer]
> I suppose we can not do use.asResource after doing await (twice).

Correct, you cannot have an `await` anywhere in the call-args of a side effect because that means the side effect will execute once the arguments finally resolve (which will then be across an async gap).

> Will be the following a good solution?

I might suggest hoisting the `use(rootBashrcTaggedFileUtility)` above all of the awaits for readability purposes (up near `getConnection`), but yes, that is likely exactly how I would implement that workaround.

**Reply by busslina** [reply]
Cool

**Reply by GregoryConrad** [reply]
Just thought of another approach:

```dart
final capsule = use(someAsyncCapsule);
return use.asResource(
  getThing: () async => await yourAsyncThing(await capsule),
);
```

Maybe that'll work based on your situation, maybe not. But it might clean things up a bit.
